File: Tests/Strata.Foundation.CoreTest/test/typescript/Utility/MappedTest.ts
import "jest"
import {Mapped,Optional,NullPointerException} from "strata.foundation.core/Utility";

describe(
    "Mapped",
    () =>
    {
        it(
            "testSuccessCreation",
            () =>
            {
                const result: Mapped<string,string> = Mapped.of("testMethod","testInput","testOutput");

                expect(result.isSuccess()).toBeTruthy();
                expect(result.isFailure()).toBeFalsy();
                expect(result.getMethod()).toBe("testMethod");
                expect(result.getInput()).toBe("testInput");
                expect(result.getOutput().isPresent()).toBeTruthy();
                expect(result.getOutput().get()).toBe("testOutput");
                expect(result.getThrowable().isEmpty()).toBeTruthy();
            });
        it(
            "testFailureCreation",
            () =>
            {
                const error: Error = new Error("test error");
                const result: Mapped<string,string> = Mapped.of<string,string>("testMethod","testInput",error);

                expect(result.isSuccess()).toBeFalsy();
                expect(result.isFailure()).toBeTruthy();
                expect(result.getMethod()).toBe("testMethod");
                expect(result.getInput()).toBe("testInput");
                expect(result.getOutput().isEmpty()).toBeTruthy();
                expect(result.getThrowable().isPresent()).toBeTruthy();
                expect(result.getThrowable().get().message).toBe("test error");
            });
        it(
            "testNullMethod",
            () =>
            {
                try
                {
                    Mapped.of(null as unknown as string,"testInput","testOutput");
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NullPointerException).toBeTruthy();
                }
            });
        it(
            "testNullInput",
            () =>
            {
                try
                {
                    Mapped.of("testMethod",null as unknown as string,"testOutput");
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NullPointerException).toBeTruthy();
                }
            });
        it(
            "testNullOutput",
            () =>
            {
                const result: Mapped<string,string> = Mapped.of("testMethod","testInput",null as unknown as string);

                expect(result.isSuccess()).toBeFalsy();
                expect(result.isFailure()).toBeTruthy();
                expect(result.getOutput().isEmpty()).toBeTruthy();
                expect(result.getThrowable().isPresent()).toBeTruthy();
            });
    });

File: Tests/Strata.Foundation.CoreTest/test/java/strata/foundation/core/utility/MappedTest.java
//////////////////////////////////////////////////////////////////////////////
// MappedTest.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

@Tag("CommitStage")
public
class MappedTest
{
    private Mapped<String,String> success;
    private Mapped<String,String> failure;

    @BeforeEach
    public void
    setUp()
    {
        success = Mapped.of("testMethod","testInput","testOutput");
        failure = Mapped.of("testMethod","testInput",new RuntimeException("test error"));
    }

    @Test
    public void
    testSuccessCreation()
    {
        assertTrue(success.isSuccess());
        assertFalse(success.isFailure());
        assertEquals("testMethod",success.getMethod());
        assertEquals("testInput",success.getInput());
        assertTrue(success.getOutput().isPresent());
        assertEquals("testOutput",success.getOutput().get());
        assertFalse(success.getException().isPresent());
    }

    @Test
    public void
    testFailureCreation()
    {
        assertFalse(failure.isSuccess());
        assertTrue(failure.isFailure());
        assertEquals("testMethod",failure.getMethod());
        assertEquals("testInput",failure.getInput());
        assertFalse(failure.getOutput().isPresent());
        assertTrue(failure.getException().isPresent());
        assertEquals("test error",failure.getException().get().getMessage());
    }

    @Test
    public void
    testNullMethod()
    {
        try
        {
            Mapped.of(null,"testInput","testOutput");
            fail("Should have thrown NullPointerException");
        }
        catch (NullPointerException e) {}
    }

    @Test
    public void
    testNullInput()
    {
        try
        {
            Mapped.of("testMethod",(String)null,"testOutput");
            fail("Should have thrown NullPointerException");
        }
        catch (NullPointerException e) {}
    }

    @Test
    public void
    testNullOutput()
    {
        Mapped<String,String> result = Mapped.of("testMethod","testInput",(String)null);

        assertFalse(result.isSuccess());
        assertTrue(result.isFailure());
        assertFalse(result.getOutput().isPresent());
        assertTrue(result.getException().isPresent());
    }

    @Test
    public void
    testNullThrowable()
    {
        Mapped<String,String> result = Mapped.of("testMethod","testInput",(Throwable)null);

        assertFalse(result.isSuccess());
        assertTrue(result.isFailure());
        assertTrue(result.getException().isPresent());
    }
}

//////////////////////////////////////////////////////////////////////////////
